File: 72_Edit Distance/72_Edit Distance.cpp
#include <iostream>
#include <cstdlib>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

//编辑距离问题属于动态规划
class Solution {
public:
	int minDistance(string word1, string word2) {
		//若其中一个字符串为空，则最短编辑距离为另一字符串的长度
		if (word1.empty())
			return word2.size();
		else if (word2.empty())
			return word1.size();

		//注意下标处理
		int m = word1.size() + 1, n = word2.size() + 1;

		//记录编辑距离的二维数组
		vector<vector<int> > distance(m, vector<int>(n, 0));
		for (int i = 0; i < m; i++)
			distance[i][0] = i;
		for (int j = 0; j < n; j++)
			distance[0][j] = j;

		for (int i = 1; i < m; i++)
		{
			for (int j = 1; j < n; j++)
			{
				if (word1[i-1] == word2[j-1])
					distance[i][j] = distance[i - 1][j - 1];
				else
					distance[i][j] = distance[i - 1][j - 1] + 1;

				distance[i][j] = min(distance[i][j], min(distance[i - 1][j] + 1, distance[i][j - 1] + 1));
			}//for
		}//for
		return distance[m - 1][n - 1];
	}
};

int main()
{
	Solution s;
	cout << s.minDistance("eeba", "abca") << endl;

	system("pause");
	return 0;
}

File: 72. Edit Distance.java
/**
 * 72. Edit Distance
 *
 * Given two words word1 and word2, find the minimum number of operations required to convert word1 to word2.
 *
 * You have the following 3 operations permitted on a word:
 *
 *     Insert a character
 *     Delete a character
 *     Replace a character
 */
public class NO_72 {
    private Solution solution = new Solution();

    public static void main(String[] args) {
        System.out.println(new NO_72().solution.minDistance("sea", "eat"));
    }

    class Solution {
        public int minDistance(String word1, String word2) {
            int m = word1.length(), n = word2.length();

            if (m * n == 0) {
                return m + n;
            }

            int[][] dp = new int[m+1][n+1];
            dp[0][0] = 0;

            for (int i=1; i<=m; ++i) {
                dp[i][0] = i;
            }

            for (int j=1; j<=n; ++j) {
                dp[0][j] = j;
            }

            for (int i=1; i<=m; ++i) {
                for (int j=1; j<=n; ++j) {
                    if (word1.charAt(i-1) == word2.charAt(j-1)) {
                        dp[i][j] = dp[i-1][j-1];
                    } else {
                        dp[i][j] = Math.min(Math.min(dp[i-1][j], dp[i][j-1]), dp[i-1][j-1]) + 1;
                    }
                }
            }

            return dp[m][n];
        }
    }
}
